In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import datetime as dt
import matplotlib.pyplot as plt
import sys
pd.set_option('display.max_columns', None)
np.set_printoptions(threshold=sys.maxsize)

In [2]:
# we can not merge step2 file with beneficary_data_tran. it is a lage data file
# This notebook, we will investigest the step 2 file

In [3]:
#global variable

step2 = 'Merge_claim_transaction.csv'
#beneficary_tran = 'beneficary_data_tran.csv'

all_claim_data_step_2  = pd.read_csv(step2)
#beneficary_tran_data = pd.read_csv(beneficary_tran) 

D:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3049: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [4]:
#key to string
#all_claim_data_step_2['policyno']  =  all_claim_data_step_2['policyno'].astype('str')
#beneficary_tran_data['policyno']  = beneficary_tran_data['policyno'].astype('str')

In [5]:
#step 3
#claim + policy_info+customer+beneficary
#all_claim_data_step_3 = all_claim_data_step_2.merge(beneficary_tran_data,  on=['policyno'], how = 'left') # represent 

In [6]:
all_claim_data_step_2.drop(columns = ['Unnamed: 0'], inplace = True)

In [7]:
all_claim_data_step_2.drop(columns = ['effectivedate_y', 'plancode_y', 'branch_y'], inplace = True)
all_claim_data_step_2.rename(columns={"effectivedate_x": "effectivedate", "plancode_x": "plancode", "branch_x":"branch"}, inplace = True)

# Date minipulation

In [8]:
#Thai date to chistien date
all_claim_data_step_2['accidentdate'] = all_claim_data_step_2['accidentdate'] - 5430000
all_claim_data_step_2['receivedate'] = all_claim_data_step_2['receivedate'] - 5430000
all_claim_data_step_2['claimokdate'] = all_claim_data_step_2['claimokdate'] - 5430000
all_claim_data_step_2['claimpaydate'] = all_claim_data_step_2['claimpaydate'] - 5430000
all_claim_data_step_2['effectivedate'] = all_claim_data_step_2['effectivedate'] - 5430000
all_claim_data_step_2['birthdate'] = all_claim_data_step_2['birthdate'] - 5430000

#### accidentdate

In [9]:
# error position
tim_temp = pd.to_datetime(all_claim_data_step_2['accidentdate'], format="%Y%m%d",errors='coerce').to_frame()
tim_temp[tim_temp.accidentdate.isnull()]

accidentdate
19563            NaT
630348           NaT
1519435          NaT
1538666          NaT
1620899          NaT
1832303          NaT
1832304          NaT
2266300          NaT

In [10]:
all_claim_data_step_2['accidentdate'].iloc[19563,] 

20170931

In [11]:
all_claim_data_step_2['accidentdate'].iloc[630348,] 

-5429982

Note position 630348, value = 18. To much error, we should drop this row.

In [12]:
all_claim_data_step_2['accidentdate'].iloc[1519435,] 

20191131

In [13]:
all_claim_data_step_2['accidentdate'].iloc[1538666,] 

170605

Note position 1538666, value = 170605. To much error, we should drop this row.

In [14]:
all_claim_data_step_2['accidentdate'].iloc[1620899,] 

20180231

In [15]:
all_claim_data_step_2['accidentdate'].iloc[1832303,] 

180908

Note position 1832303, value = 180908. To much error, we should drop this row.

In [16]:
all_claim_data_step_2['accidentdate'].iloc[1832304,] 

180908

In [17]:
all_claim_data_step_2['accidentdate'].iloc[2266300,]

20191131

In [18]:
all_claim_data_step_2['accidentdate'] = all_claim_data_step_2['accidentdate'].astype('str')
all_claim_data_step_2['accidentdate'] = all_claim_data_step_2['accidentdate'].replace({'20170931':'20170930', '20191131':'20191130', '20180231':'20180228', '20191131':'20191130'})
all_claim_data_step_2 = all_claim_data_step_2.drop([630348, 1538666,1832303,1832304], axis=0)

In [19]:
all_claim_data_step_2['accidentdate'] = pd.to_datetime(all_claim_data_step_2['accidentdate'], format="%Y%m%d")

### receivedate

In [20]:
tim_temp = pd.to_datetime(all_claim_data_step_2['receivedate'], format="%Y%m%d",errors='coerce').to_frame()
tim_temp[tim_temp.receivedate.isnull()]

Empty DataFrame
Columns: [receivedate]
Index: []

NANI! NO error.

In [21]:
all_claim_data_step_2['receivedate'] = pd.to_datetime(all_claim_data_step_2['receivedate'], format="%Y%m%d")

### claimokdate and claimpaydate

In [22]:
tim_temp = pd.to_datetime(all_claim_data_step_2['claimokdate'], format="%Y%m%d",errors='coerce').to_frame()
tim_temp[tim_temp.claimokdate.isnull()]

claimokdate
224             NaT
244             NaT
630             NaT
634             NaT
840             NaT
1023            NaT
1295            NaT
1515            NaT
1555            NaT
1826            NaT
2189            NaT
2473            NaT
2705            NaT
2931            NaT
3620            NaT
4178            NaT
4262            NaT
4379            NaT
4517            NaT
4518            NaT
4805            NaT
4881            NaT
4927            NaT
4989            NaT
5005            NaT
5301            NaT
5389            NaT
5462            NaT
5517            NaT
5543            NaT
...             ...
2929623         NaT
2930535         NaT
2930820         NaT
2930961         NaT
2931155         NaT
2931272         NaT
2931307         NaT
2931308         NaT
2931318         NaT
2931819         NaT
2932501         NaT
2933013         NaT
2933052         NaT
2933082         NaT
2933535         NaT
2933628         NaT
2933652         NaT
2933836         NaT
2933845         NaT
2933901         NaT
2934019         NaT
2934290         NaT
2934319         NaT
2934350         NaT
2934382         NaT
2934534         NaT
2934552         NaT
2934695         NaT
2934779         NaT
2934826         NaT

[14980 rows x 1 columns]

In [23]:
tim_temp = pd.to_datetime(all_claim_data_step_2['claimpaydate'], format="%Y%m%d",errors='coerce').to_frame()
tim_temp[tim_temp.claimpaydate.isnull()]

claimpaydate
11               NaT
12               NaT
34               NaT
43               NaT
60               NaT
92               NaT
111              NaT
112              NaT
114              NaT
122              NaT
129              NaT
131              NaT
146              NaT
147              NaT
173              NaT
178              NaT
181              NaT
187              NaT
189              NaT
190              NaT
199              NaT
204              NaT
224              NaT
244              NaT
254              NaT
266              NaT
267              NaT
270              NaT
302              NaT
306              NaT
...              ...
2934552          NaT
2934554          NaT
2934579          NaT
2934587          NaT
2934604          NaT
2934613          NaT
2934626          NaT
2934636          NaT
2934646          NaT
2934662          NaT
2934671          NaT
2934674          NaT
2934683          NaT
2934686          NaT
2934695          NaT
2934704          NaT
2934710          NaT
2934715          NaT
2934720          NaT
2934746          NaT
2934763          NaT
2934771          NaT
2934773          NaT
2934776          NaT
2934779          NaT
2934786          NaT
2934821          NaT
2934826          NaT
2934835          NaT
2934836          NaT

[303985 rows x 1 columns]

__Note__: We are not serious with claimokdate and claimpaydate. __But why they don't have same row.__

In [24]:
all_claim_data_step_2['claimokdate'] = pd.to_datetime(all_claim_data_step_2['claimokdate'], format="%Y%m%d", errors='coerce')
all_claim_data_step_2['claimpaydate'] = pd.to_datetime(all_claim_data_step_2['claimpaydate'], format="%Y%m%d", errors='coerce')

### effectivedate

In [25]:
all_claim_data_step_2['effectivedate'] = all_claim_data_step_2['effectivedate'].astype('str')
tim_temp = pd.to_datetime(all_claim_data_step_2['effectivedate'], format="%Y%m%d",errors='coerce').to_frame()
tim_temp[tim_temp.effectivedate.isnull()]

effectivedate
124269            NaT
535622            NaT
988266            NaT
1913753           NaT
2309638           NaT
2431424           NaT
2473079           NaT

In [26]:
all_claim_data_step_2['effectivedate'].iloc[124269,] 

'-5430000'

In [27]:
all_claim_data_step_2['effectivedate'].iloc[535622,] 

'-5430000'

In [28]:
all_claim_data_step_2['effectivedate'].iloc[988266,] 

'20070524'

In [29]:
all_claim_data_step_2['effectivedate'].iloc[1913753,] 

'20120801'

In [30]:
all_claim_data_step_2['effectivedate'].iloc[2309638,] 

'20171028'

In [31]:
all_claim_data_step_2['effectivedate'].iloc[2431424,] 

'19980428'

In [32]:
all_claim_data_step_2['effectivedate'].iloc[2473079,] 

'20020524'

In [33]:
all_claim_data_step_2['effectivedate'] = pd.to_datetime(all_claim_data_step_2['effectivedate'], format="%Y%m%d", errors='coerce')

### birthdate

In [34]:
all_claim_data_step_2['birthdate'] = all_claim_data_step_2['birthdate'].astype('str')
all_claim_data_step_2['birthdate'] = pd.to_datetime(all_claim_data_step_2['birthdate'], format="%Y%m%d", errors='coerce')

# Tryping

In [35]:
all_claim_data_step_2

policyno claimtype accidentdate receivedate sectioncode   orderno  \
0        82115618         A   2013-01-09  2013-01-16          02  56W05116   
1        35570103         A   2013-03-01  2013-03-08          94  56A09505   
2        35570103         A   2013-03-01  2013-03-08          94  56A09505   
3        23056157         A   2019-10-02  2019-10-03          02  62F25090   
4        32867270         A   2016-09-05  2016-10-11          16  59A11942   
5        32528743         A   2012-10-22  2012-10-29          92  55A40431   
6        32528743         A   2012-10-22  2012-10-29          92  55A40431   
7        36131650         A   2014-11-28  2014-12-17          10  57A20363   
8        36844161         A   2018-02-19  2018-03-14          19  61A03643   
9        33271157         A   2012-12-06  2012-12-06          02  55X04823   
10       21727035         A   2012-02-24  2012-03-05          92  55A08964   
11       20770504         A   2019-11-30  2019-12-28          94  62A62638   
12       20770504         A   2019-11-30  2019-12-28          94  62A62638   
13       82138446         A   2012-11-18  2012-12-18          02  55X08851   
14       20443147         A   2011-03-03  2011-03-08          07  54A02999   
15       34719068         A   2013-06-03  2013-06-05          02  56C44005   
16       21880613         A   2014-09-01  2014-09-03          02  57C82726   
17       34649437         A   2010-12-31  2011-02-22          11  54A01592   
18       20689773         A   2011-08-21  2011-08-22          13  54A11940   
19       22374398         A   2018-11-30  2018-11-30          02  61Y58874   
20       30936335         A   2013-09-25  2013-10-15          10  56A17821   
21       82065607         A   2018-12-12  2018-12-13          02  61Y67729   
22       21959529         D   2013-05-14  2013-05-17          D3  56A00085   
23       22652588         A   2016-08-20  2016-08-20          02  59X34680   
24       32963163         A   2016-05-24  2016-05-26          02  59C53408   
25       32963163         A   2016-05-24  2016-05-26          02  59C53408   
26       35720190         A   2014-08-04  2014-09-24          16  57A12049   
27       35185883         A   2012-10-07  2012-10-08          02  55W85589   
28       22149679         A   2013-03-09  2013-03-09          02  56W24638   
29       34574861         A   2012-07-28  2012-08-27          02  55C67236   
...           ...       ...          ...         ...         ...       ...   
2934829  35570303         A   2013-07-26  2013-08-13          16  56A10918   
2934830  37071666         A   2018-09-19  2018-09-24          08  61A18050   
2934831  36373193         A   2015-07-12  2015-07-15          08  58A12172   
2934832  34838367         A   2016-06-07  2016-11-16          15  59A13894   
2934833  34838367         A   2016-06-07  2016-11-16          15  59A13894   
2934834  36163690         A   2015-12-15  2015-12-21          14  58A22313   
2934835  36466671         A   2016-01-08  2016-01-08          02  59F00439   
2934836  36466671         A   2016-01-08  2016-01-08          02  59F00439   
2934837  21586728         A   2011-09-25  2011-09-30          02  54C70959   
2934838  35395587         A   2019-01-16  2019-01-17          94  62C07597   
2934839  34955861         A   2017-09-28  2017-09-28          02  60X76689   
2934840  31448941         A   2018-12-26  2019-01-15          02  62W10653   
2934841  31448941         A   2018-12-26  2019-01-15          02  62W10653   
2934842  21274984         A   2017-05-24  2017-06-12          02  60A21280   
2934843  22371045         A   2014-12-29  2014-12-31          02  57F27268   
2934844  36276419         A   2017-07-03  2017-07-17          06  60A16054   
2934845  36678332         A   2017-07-27  2017-08-04          50  60A05508   
2934846  32209302         A   2013-05-03  2013-05-23          07  56A07430   
2934847  80080467         A   2017-05-05  2017-05-08          02  60W75100   
2934848  21198288         A   2

In [36]:
#list of all columns
All_col_name = list(all_claim_data_step_2.columns)
id_col_name = ['policyno', 'unitcode', 'orderno', 'considercode', 'sourceyymm', 'salesid', 'personid', 'referenceid', 'branchconsiderid']
date_col_name = ['accidentdate', 'receivedate', 'claimokdate','claimpaydate', 'effectivedate', 'admitstartdate','admitenddate', 'birthdate']
category_col_name = ['claimtype', 'sectioncode', 'polyear', 'plancode', 'ridertype', 'src_filename', 'groupname', 'branch', 'lcode', 'claimpaytype', 'policystatus1', 'policystatus2', 'causecode1', 'causedetail1', 'causecode2', 'causedetail2', 'hospitalcode', 'provincecode', 'psysiciancertificate', 'admitflag', 'admittype', 'remarkno', 'remarkdesc', 'sourceyymm', 'claimchannel', 'illnesscode', 'illnessdetail', 'contclaimflag', 'msttype', 'contractexcludeflag', 'payerflag', 'payeridno', 'occupationcode', 'branch', 'sex', 'flagfraud', 'nation', 'marriage', 'bankruptcy_flag']
numerical_col_name = ['claimamount', 'tdday', 'pdday', 'admitday', 'sfamount', 'sfrate', 'ansamount', 'oramount', 'deductamount', 'sum', 'insuredage']

In [37]:
for col in id_col_name:
    all_claim_data_step_2[col] = all_claim_data_step_2[col].astype('str')

for col in date_col_name:
    all_claim_data_step_2[col] = pd.to_datetime(all_claim_data_step_2[col], format="%Y%m%d", errors='coerce')
    
for col in category_col_name:
     all_claim_data_step_2[col] = all_claim_data_step_2[col].astype('category')

In [38]:
all_claim_data_step_2.sort_values(['personid','accidentdate'])

policyno claimtype accidentdate receivedate sectioncode   orderno  \
884133   21537982         A   2011-01-19  2011-01-21          02  54C04721   
1569227  21537982         A   2012-04-04  2012-04-05          02  55W26023   
562788   21537982         A   2012-12-19  2012-12-22          02  55D02527   
1107690  21537982         A   2014-12-16  2014-12-26          10  57A20976   
2896073  21537982         A   2014-12-28  2015-01-12          10  58A00500   
1916031  25156448         A   2015-06-26  2015-07-09          91  58A26352   
649043   32925169         A   2011-04-06  2011-07-27          14  54A09427   
340943   22355009         A   2015-11-29  2016-10-19          93  59A40314   
340944   22355009         A   2015-11-29  2016-10-19          93  59A40314   
828041   22355009         A   2018-05-13  2018-07-02          02  61C79289   
1194597  22355009         A   2018-11-27  2018-11-28          02  61Y57491   
1194598  22355009         A   2018-11-27  2018-11-28          02  61Y57491   
2207834  22355009         A   2018-11-27  2018-12-13          02  61Y67751   
2207835  22355009         A   2018-11-27  2018-12-13          02  61Y67751   
2345507  22355009         A   2018-11-27  2019-03-06          02  62C32265   
2853117  22355009         A   2018-11-27  2019-01-12          02  62W08053   
2853118  22355009         A   2018-11-27  2019-01-12          02  62W08053   
2202885  20615234         A   2012-06-10  2012-07-18          04  55A06609   
1098410  80047876         A   2013-04-01  2013-04-19          04  56A03350   
1451442  80047876         A   2013-04-01  2013-04-01          02  56W33967   
620392   21310628         D   2016-01-08  2016-01-13          04  59A00304   
2735617  20711411         A   2013-05-18  2013-05-28          92  56A20110   
2735618  20711411         A   2013-05-18  2013-05-28          92  56A20110   
1182693  20711411         A   2015-06-06  2015-06-24          92  58A24287   
1182694  20711411         A   2015-06-06  2015-06-24          92  58A24287   
2596062  20711411         A   2016-06-29  2016-09-08          92  59A34423   
2596063  20711411         A   2016-06-29  2016-09-08          92  59A34423   
2152259  21552320         A   2011-10-25  2011-11-23          18  54A11899   
2661853  34942996         A   2015-10-31  2015-10-31          02  58X51509   
404917   25194090         A   2011-04-21  2011-07-12          04  54A07192   
...           ...       ...          ...         ...         ...       ...   
1088051   9066565         D   2018-07-14  2018-07-17          E3  61A00014   
351215    6700756         A   2018-07-16  2018-10-12          91  61A50272   
2000634   6007870         D   2018-08-16  2019-05-29          94  62A25818   
1955835   9106092         D   2018-08-20  2018-09-18          94  61A45193   
2823143   9053672         D   2018-08-26  2018-09-03          FN  61A00019   
2864518   6005906         A   2018-10-30  2018-11-20          02  61D46081   
1107487   9094887         D   2018-11-01  2018-11-05          10  61A18978   
2753311   9077739         D   2018-11-09  2018-11-27          16  61A12505   
768404    9119309         D   2018-12-09  2018-12-13          29  61A00071   
2826684   9105978         D   2018-12-11  2018-12-20          D9  61A00218   
2443628   9092907         D   2018-12-30  2019-01-09          13  62A00436   
316919    9098920         D   2019-01-19  2019-04-17          15  62A04649   
327619    9080133         D   2019-02-06  2019-02-13          AZ  62A00016   
1723672   6006609         D   2019-02-25  2019-03-06          94  62A11532   
2815724   6005374         A   2019-03-19  2019-03-20          06  62A06534   
1491067   9127411         D   2019-04-10  2019-04-17          AR  62A00027   
41936     9122265         D   2019-04-19  2019-05-29          91  62A25725   
1307783   9094861         D   2019-04-22  2019-04-24          CF  62A00010   
1552465   9112847         D   2019-05-26  2019-05-29          AB  62A00004   
2005257   9098051         D   2

In [39]:
all_claim_data_step_2.ridertype.unique()

[ME, NaN, S25, วพ5, อ02, ..., D03, T50, S80, AD, อพ5]
Length: 145
Categories (144, object): [ME, S25, วพ5, อ02, ..., T50, S80, AD, อพ5]

In [40]:
df = all_claim_data_step_2[['personid','accidentdate','ridertype']]
df_pivot = df.pivot_table(index= ['personid','accidentdate'], columns='ridertype', aggfunc=len, fill_value=0)

In [41]:
df_pivot.to_csv('pivot_ridertype_per_claim_count.csv')

In [42]:
all_claim_data_step_2.to_csv('Clearn_merge_claim_transaction.csv')